### PR TITLE
cloudprober 0.12.0 (new formula)

### PR DESCRIPTION
**Pull request by manugarg**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/cloudprober.rb
Comment (SMillerDev) [inline Formula/cloudprober.rb:16]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Comment (manugarg) [reply Formula/cloudprober.rb]: Done.

File: Formula/cloudprober.rb
Comment (SMillerDev) [inline Formula/cloudprober.rb:39]: Doesn't this work?
```suggestion
    out = shell_output(bin/"cloudprober --logtostderr 2>&1")
    assert_match "Initialized status surfacer", out
  end
```

Comment (manugarg) [reply Formula/cloudprober.rb:39]: Unfortunately no, because Cloudprober just keeps running indefinitely.

**Comment by manugarg** [comment]
@SMillerDev I was wondering if you any further comments / questions?

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [inline Formula/cloudprober.rb:22]:
```suggestion
    r, w, pid, = PTY.spawn(bin/"cloudprober --logtostderr 2>&1")
```
but please use longer variables as I have no idea what these mean. Same for `li`, `r` below.

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [inline Formula/cloudprober.rb:12]: Is there a `make install` we can use?

Comment (dawidd6) [reply Formula/cloudprober.rb:12]: There is, it's just running `go install` under the hood.

Comment (MikeMcQuaid) [reply Formula/cloudprober.rb:12]: Yeh, let's use that then for future-proofing if they install more in future.

Comment (manugarg) [reply Formula/cloudprober.rb:12]: I'd argue this is better. make install doesn't take the prefix, while bin.install will do the right thing for brew. No?

File: Formula/cloudprober.rb
Comment (dawidd6) [inline Formula/cloudprober.rb:29]: I think we need to raise an error if line wasn't found. Otherwise brew will think the test block completed successfully. If I read the code correctly, just returning a false-y value will not trigger a failure.

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [reply Formula/cloudprober.rb:12]: `make install` should be able to take a prefix and is more future proof if in future you ever decide to install e.g. docs.

I didn't realise you were the upstream for this? We frown upon projects submitting themselves to Homebrew unless they can e.g. point to user requests for this. Do you have any issues etc. for this?

Comment (manugarg) [reply Formula/cloudprober.rb:12]: > I didn't realise you were the upstream for this? We frown upon projects submitting themselves to Homebrew unless they can e.g. point to user requests for this. Do you have any issues etc. for this?

Actually, I am also the user here. I use brew myself and thought it will be easier for me (and Cloudprober users) if brew could install cloudprober. Also my another open-source project, pacparser, was already on brew, so I thought to give it a try.

I see your point regarding users requesting for it, but as a user I don't file a bug if a software is not available on brew. I just find it somewhere else. So by that criteria, you'll be missing a lot of genuine users.

In case, you're concerned about cloudprober not being worthy, I think Cloudprober has a significant user base (2.5M docker pulls, 1.4k stars on [the original project](https://github.com/google/cloudprober) (I had to move the development after I left the project) and users will benefit from it being on brew.

Comment (MikeMcQuaid) [reply Formula/cloudprober.rb:12]: Thanks for explaining @manugarg! Talked with some other maintainers and I'm 👍🏻 on including this.

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [inline Formula/cloudprober.rb:36]: Can you see if either of these would work?

```suggestion
  test do
    io = IO.popen("#{bin}/cloudprober", "--logtostderr", err: [:child, :out])
    io.any? do |line|
      /Initialized status surfacer/.match?(line)
    end
  end
```

```suggestion
  test do
    require "open3"

    Open3.popen3("#{bin}/cloudprober --logtostderr 2>&1") do |_, stdout|
      stdout.any? do |line|
        /Initialized status surfacer/.match?(line)
      end
    end
  end
```

As-is the test is a bit unwieldy. Things that may help frame the above:
- no `require` is preferred if possible
- I'm unsure if `Initialized status surfacer` is the first line output/last line output/etc. If it's the first line output: the above can probably be simplified
- our `test do` already should handle timeouts, closing and killing processes. Please check and see if I'm wrong here and only include them if required

Thanks again!

Comment (manugarg) [reply Formula/cloudprober.rb]: (Looks like my earlier message got lost.)

Thanks a lot for the suggestions. I verified that first suggestion works.

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [inline Formula/cloudprober.rb:19]: @manugarg last question: is this always the first line printed? If so, we can simplify further. If not: no worries and will approve as-is.

File: Formula/cloudprober.rb
Comment (MikeMcQuaid) [inline Formula/cloudprober.rb:16]:
```suggestion
    io = IO.popen("#{bin}/cloudprober", "--logtostderr", err: [:child, :out])
```

again, assuming it works, is a little more elegant

Comment (manugarg) [reply Formula/cloudprober.rb:16]: I did try this. It didn't work, that's why I made it one string.

File: Formula/cloudprober.rb
Comment (manugarg) [reply Formula/cloudprober.rb:19]: Sorry, missed responding to that. Yeah, it's not the first line. It gets printed after cloudprober initializes.

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution and for handling all the back and forth! Without people like you submitting PRs we couldn't run this project. You rock, @manugarg!

**Comment by MikeMcQuaid** [comment]
(cancelled this, published prematurely, need to wait on CI!)

**Comment by manugarg** [comment]
@MikeMcQuaid it may require re-running the CI jobs, 3 jobs could not be scheduled and finally got canceled.

**Pull request closed without merging**
